[run-webkit-tests] Run largest shards first

JonWBedard opened the pull request:
#### 7c24f5486272e5ec5c88a01e5959d691c3a28f59
<pre>
[run-webkit-tests] Run largest shards first
<a href="https://bugs.webkit.org/show_bug.cgi?id=228429">https://bugs.webkit.org/show_bug.cgi?id=228429</a>
rdar://81184667

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(TestShard.__init__): Estimate runtime of each shard.
(TestShard.__getstate__): Pass shard_time_estimate between processes.
(TestShard.__setstate__): Ditto.
(Sharder._shard_by_directory):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner_unittest.py:
(SharderTests.test_shard_by_dir):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c24f5486272e5ec5c88a01e5959d691c3a28f59

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6336 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4949 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5188 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4903 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5501 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6352 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2981 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9288 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4012 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4337 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5709 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5971 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4562 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4799 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3910 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4919 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4306 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1203 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8369 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5052 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4668 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1316 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -499,12 +499,75 @@ def _run_single_test(self, driver, test_input, stop_when_done):
+        'imported/w3c/web-platform-tests/xhr': 1002,
+        'jquery': 842,
+        'media': 436,
+        'platform': 804,
```
This is nonsense, including all of `platform` as one line?

JonWBedard replied:
It's not number of tests, it's average test time (so tests in "platform", on average, take 804 milliseconds to complete)

gsnedders replied:
But does it make sense for `platform` to be considered a category here? It seems a bit weird to have it as different to the overall default. And, like, do we really believe that `platform/mac` is representative of `platform/ios`?

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -499,12 +499,75 @@ def _run_single_test(self, driver, test_input, stop_when_done):
     """A test shard is a named list of TestInputs."""
+    DEFAULT_WEIGHT = 256
+    WEIGHT_ESTIMATES = {
+        'animations': 438,
```
Do we want the (presumably exact) current number, or do we want to approximate it and round it (to multiples of 256, to powers of 2, etc.)?

JonWBedard replied:
Prefacing this comment by pointing out that this change was originally just meant as an experiment (which is why it's still marked as a "draft"). That being said, the "experiment" was surprisingly successful (10-15% runtime improvement, no side effects), so it seems likely that a version of this change will find it's way up for review in short order.

These numbers are the average time, in milliseconds, that each individual test takes to complete. The average for the whole suite is around 200. The thinking here is that we can get a pretty accurate estimate of shard time as long as our guess for how long tests in a given shard will take isn't orders of magnitude off. I actually calculated average tests time for every shard, but if a child's average test time was within 25% of its parent's, I deleted the child. I think did a bit of manual curation, deleting some measurements if there weren't very many tests in a shard.

gsnedders replied:
I think I'm still in favour of rounding this further than the current raw figures; yes, it'll make it less accurate, but I suspect we have more precision here than the noise really allows us to justify.

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -499,12 +499,75 @@ def _run_single_test(self, driver, test_input, stop_when_done):
+            key = '{}/{}'.format(key, part)
+            weight = self.WEIGHT_ESTIMATES.get(key, weight)
+
+        self.cost = len(self.test_inputs) * weight
```
Does this not end up with the weight of those which we explicitly have estimates for being n*2 for n tests?

JonWBedard replied:
Weight estimates aren't number of tests, they're average number of milliseconds per test. The "cost" of a shard is basically estimated runtime (which is probably a better name than cost), and is computed by our guess for the average runtime of a test in that shard multiplied by the number of tests in the shard.

JonWBedard committed:
[run-webkit-tests] Run largest shards first
https://bugs.webkit.org/show_bug.cgi?id=228429
rdar://81184667

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(TestShard.__init__): Estimate runtime of each shard.
(TestShard.__getstate__): Pass shard_time_estimate between processes.
(TestShard.__setstate__): Ditto.
(Sharder._shard_by_directory):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner_unittest.py:
(SharderTests.test_shard_by_dir):

gsnedders commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -499,12 +499,75 @@ def _run_single_test(self, driver, test_input, stop_when_done):
 class TestShard(object):
     """A test shard is a named list of TestInputs."""
+    DEFAULT_AVERAGE_TEST_TIME = 256
+    AVERAGE_TEST_TIME_ESTIMATES = {
```
Looking at this again, I'm somewhat doubting whether this is the right place to actually store all this data. It's much more like configuration, rather than something tied to the runner?

JonWBedard replied:
Our test-runner is very coupled to our tests. If we wanted to put this somewhere else, it's not clear to me where it would be. I also can't think of a circumstance where we would need a second configuration like this. This `TestShard` object is, specifically, a shard of layout tests.
